stackit/internal/services/ske/cluster/resource.go

```diff
@@ -848,7 +919,144 @@ func toNodepoolsPayload(ctx context.Context, m *Model) ([]ske.Nodepool, error) {
+	for _, machine := range availableImages {
+		if machine.Name != nil && *machine.Name == osName && machine.Versions != nil {
+			availableMachineVersions = *machine.Versions
+		}
+	}
+
+	if len(availableImages) == 0 {
+		return nil, false, fmt.Errorf("there are no available machine versions for the provided machine image name %s", osName)
+	}
+
+	if versionMin == nil {
+		// Different machine OSes have different versions.
+		// If the current machine image is nil or the machine image name has been updated,
+		// retrieve the latest supported version. Otherwise, use the current machine version.
+		if currentImage == nil || currentImage.Name == nil || *currentImage.Name != osName {
+			latestVersion, err := getLatestSupportedMachineVersion(availableMachineVersions)
+			if err != nil {
+				return nil, false, fmt.Errorf("get latest supported machine image version: %w", err)
+			}
+			return latestVersion, false, nil
+		}
+		versionMin = currentImage.Version
+	} else if currentImage != nil && currentImage.Name != nil && *currentImage.Name == osName {
+		// If the os_version_min is set but is lower than the current version used in the cluster,
+		// retain the current version to avoid downgrading.
+		minimumVersion := "v" + *versionMin
+		currentVersion := "v" + *currentImage.Version
+
+		if semver.Compare(minimumVersion, currentVersion) == -1 {
+			versionMin = currentImage.Version
+		}
+	}
+
+	var fullVersion bool
+	versionExp := validate.FullVersionRegex
+	versionRegex := regexp.MustCompile(versionExp)
+	if versionRegex.MatchString(*versionMin) {
+		fullVersion = true
+	}
+
+	providedVersionPrefixed := "v" + *versionMin
+
+	if !semver.IsValid(providedVersionPrefixed) {
+		return nil, false, fmt.Errorf("provided version is invalid")
+	}
+
+	var versionUsed *string
+	var state *string
+	var availableVersionsArray []string
+	// Get the higher available version that matches the major, minor and patch version provided by the user
+	for _, v := range availableMachineVersions {
+		if v.State == nil || v.Version == nil {
+			continue
+		}
+		availableVersionsArray = append(availableVersionsArray, *v.Version)
+		vPreffixed := "v" + *v.Version
+
+		if fullVersion {
+			// [MAJOR].[MINOR].[PATCH] version provided, match available version
+			if semver.Compare(vPreffixed, providedVersionPrefixed) == 0 {
+				versionUsed = v.Version
+				state = v.State
+				break
+			}
+		} else {
+			// [MAJOR].[MINOR] version provided, get the latest patch version
+			if semver.MajorMinor(vPreffixed) == semver.MajorMinor(providedVersionPrefixed) &&
+				(semver.Compare(vPreffixed, providedVersionPrefixed) == 1 || semver.Compare(vPreffixed, providedVersionPrefixed) == 0) &&
+				(v.State != nil && *v.State != VersionStatePreview) {
+				versionUsed = v.Version
+				state = v.State
+			}
+		}
+	}
+
+	if versionUsed != nil {
+		if strings.EqualFold(*state, VersionStateDeprecated) {
+			deprecated = true
+		} else {
+			deprecated = false
+		}
```

Comment:
```suggestion
		deprecated = strings.EqualFold(*state, VersionStateDeprecated)
```

Reply:
Is simpler but maybe we lose some clarity. Anyway, if you keep the `if`, the `else` clause should be removed, deprecated is already false.